### PR TITLE
feat: argmax proving

### PR DESCRIPTION
**Pull request by nicholas-mainardi**

### CODE REVIEW

File: zkml/src/layers/transformer/logits.rs [hunk 1]
```diff
@@ -1,81 +1,740 @@
...
-                        let rows = input
+                        let (indices, maximums): (Vec<_>, Vec<_>) = input
                             .slice_last_dim()
                             .map(|row| {
-                                Tensor::new(
-                                    // we want to stack along a new dimension so we unsqueeze manually here
-                                    Shape::new(vec![1, 1]),
-                                    vec![N::from_usize(argmax_slice(row).unwrap())],
-                                )
+                                let argmax = argmax_slice(row).unwrap();
```
Comment (Zyouell) [inline zkml/src/layers/transformer/logits.rs:107]: Is it worth making `argmax_slice()` return the tuple rather than just the index (currently it appears to find the index of the maximum as well as the maximum value but then only returns the index)?

File: zkml/src/layers/transformer/logits.rs [hunk 2]
```diff
@@ -1,81 +1,740 @@
...
+    fn quantize_op<S: ScalingStrategy>(
+        self,
+        data: &S::AuxData,
+        node_id: NodeId,
+        input_scaling: &[ScalingFactor],
+    ) -> anyhow::Result<QuantizeOutput<Self::QuantizedOp>> {
+        // no need to quantize, we just propagate the scaling factors
+        let num_inputs = input_scaling.len();
+        let num_outputs = self.num_outputs(num_inputs);
+        let output_scalings = S::scaling_factors_for_node(data, node_id, num_outputs);
```
Comment (Zyouell) [inline zkml/src/layers/transformer/logits.rs:256]: Based on the current state of the code I think `output_scalings` could end up being different to `input_scalings` (as it will be calculated on fewer values) so I think we should just set `output_scalings` to be equal to `input_scalings`.

File: zkml/src/layers/transformer/logits.rs [hunk 3]
```diff
@@ -1,81 +1,740 @@
...
+}
+
+impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> ProvableOp<E, PCS> for Logits {
+    type Ctx = LogitsCtx<E>;
+
+    fn prove<T: transcript::Transcript<E>>(
+        &self,
+        node_id: NodeId,
+        _ctx: &Self::Ctx,
+        _last_claims: Vec<&Claim<E>>,
```
Comment (Zyouell) [inline zkml/src/layers/transformer/logits.rs:300]: Do we not have to prove that some claim about the output is linked to the proving we do here?

**Review by Zyouell — CHANGES_REQUESTED** [review]
A couple things, mostly that I think we can get rid of a whole sumcheck and send less data in the proof for the step. Other than that good job!

File: zkml/src/layers/transformer/logits.rs (hunk 3 above)
Comment (nicholas-mainardi) [reply zkml/src/layers/transformer/logits.rs:302]: I changed the verifier code (adding the `compute_model_output_claims` method to `VerifiableCtx` trait) to avoid computing any output claim for this layer; indeed, a claim about the one-hot encoded output will be derived from the proof of the layer, and verified against the actual output tensor.

File: zkml/src/layers/transformer/logits.rs [hunk 4]
```diff
@@ -1,81 +1,740 @@
...
+        input_shapes
+            .iter()
+            .map(|s| vec![s.dim(0), 1].into())
+            .collect()
+    }
+
+    fn split_claim_point<E: ExtensionField>(
+        point: &[E],
+        num_row_vars: usize,
+    ) -> anyhow::Result<(&[E], &[E])> {
```
[the remark this answers is not in the document]
Comment (nicholas-mainardi) [reply zkml/src/layers/transformer/logits.rs:139]: I used a `Result` as an output point as in future we might need to add some assertions that may make the function failing; shouldn't be a big issue to already return a `Result` I think but I can remove it if you prefer, it's not a big deal

File: zkml/src/layers/transformer/logits.rs [hunk 5]
```diff
@@ -1,81 +1,740 @@
...
+        let input_claim = Claim::new(diff_claim.point.clone(), max_eval - diff_claim.eval);
+
+        // build the reduced MLE of the matrix `M`, which has the same shape of `input`, from `max_values` and `output`.
+        // More specifically, `M[i][j] = max_values[i] if j == output[i], 0 otherwise`. The MLE has the row variables
+        // already fixed to `row_point`, to be later employed in the sum-check
+        let input_shape = input.get_shape();
+
+        let beta_vec = compute_betas_eval(row_point);
+        let mut reduced_m = vec![E::ZERO; input_shape.dim(input_shape.rank() - 1)];
+        beta_vec.iter().enumerate().try_for_each(|(i, beta)| {
```
[the remark this answers is not in the document]
Comment (nicholas-mainardi) [reply zkml/src/layers/transformer/logits.rs]: Done in commit https://github.com/Lagrange-Labs/deep-prove/pull/228/commits/9c4212a01fa0531aa993eef37e8b2bf7c0cd6f10

File: zkml/src/layers/transformer/logits.rs [hunk 6]
```diff
@@ -1,81 +1,740 @@
...
+        let final_input_claim = same_poly_proof.extract_claim();
+
+        let proof = LogitsProof {
+            logup_proof,
+            max_eval,
+            max_commitment,
+            max_mat_eval: max_matrix_eval,
+            sumcheck_proof,
+            hadamard_proof,
+            hadamard_evals: state.get_mle_final_evaluations()[..2].to_vec(),
```
[the remark this answers is not in the document]
Comment (nicholas-mainardi) [reply zkml/src/layers/transformer/logits.rs]: Done in commit https://github.com/Lagrange-Labs/deep-prove/pull/228/commits/9c4212a01fa0531aa993eef37e8b2bf7c0cd6f10
